File: vllm_doctor/cli/history.py
import json as json_mod
from datetime import datetime
from pathlib import Path

import typer
from rich.box import SIMPLE
from rich.console import Console
from rich.table import Table
from rich.text import Text

from vllm_doctor.cli import Format, app
from vllm_doctor.config import load_config
from vllm_doctor.reports import json as json_report
from vllm_doctor.reports import text as text_report
from vllm_doctor.reports.text import HEALTH_COLOR
from vllm_doctor.stores import HistoryStore

history_app = typer.Typer(help="Local diagnosis history commands")
app.add_typer(history_app, name="history")


@history_app.command()
def list(
    output: Format = typer.Option(Format.text, "--output", "-o", help="Output format (text or json)."),
    verbose: bool = typer.Option(False, "--verbose", "-v", help="Show additional columns."),
    config_path: Path | None = typer.Option(
        None, "--config", "-c", help="Path to config file (default: vllm-doctor.toml)."
    ),
) -> None:
    config = load_config(config_path)
    with HistoryStore(config.database.url) as store:
        runs = store.list()

    if output == Format.json:
        data = [run.model_dump() for run in runs]
        typer.echo(json_mod.dumps(data))
        return

    if not runs:
        typer.echo("No saved diagnosis runs found.")
        return

    table = Table(show_header=True, box=SIMPLE, padding=(0, 1), header_style="bold")
    table.add_column("Run ID", no_wrap=True, min_width=20)
    table.add_column("Time", no_wrap=True, min_width=16)
    table.add_column("Model", no_wrap=True, min_width=20)
    if verbose:
        table.add_column("Mode", no_wrap=True, min_width=10)
    table.add_column("Health", no_wrap=True, min_width=8)
    table.add_column("Fired", justify="right", no_wrap=True, min_width=5)

    for run in runs:
        color = HEALTH_COLOR.get(run.health, "white")
        saved = datetime.fromisoformat(run.saved_at).strftime("%Y-%m-%d %H:%M")
        row = [
            run.run_id,
            saved,
            run.model_name or "—",
        ]
        if verbose:
            row.append(run.client_mode.value)
        row += [
            Text(run.health.value, style=f"bold {color}"),
            str(run.fired_count),
        ]
        table.add_row(*row)

    console = Console()
    console.print(table)


@history_app.command()
def show(
    run_id: str = typer.Argument(..., help="ID of the saved diagnosis run."),
    output: Format = typer.Option(Format.text, "--output", "-o", help="Output format (text or json)."),
    verbose: bool = typer.Option(False, "--verbose", "-v", help="Show additional diagnostic detail."),
    config_path: Path | None = typer.Option(
        None, "--config", "-c", help="Path to config file (default: vllm-doctor.toml)."
    ),
) -> None:
    config = load_config(config_path)
    with HistoryStore(config.database.url) as store:
        result = store.get(run_id)

    if result is None:
        typer.secho(f"Error: run {run_id} not found.", fg=typer.colors.RED, err=True)
        raise typer.Exit(code=1)

    if output == Format.json:
        typer.echo(json_report.render(result, verbose=verbose))
    else:
        text_report.render(result, verbose=verbose)

File: src/cli/commands/history.rs
use std::io::IsTerminal;
use std::path::Path;

use crate::cli::args::{Format, HistoryCommand};
use crate::cli::reports::{Report, json, text};
use crate::cli::stores::{HistoryStore, RunSummary, SqliteHistoryStore};
use crate::core::models::Health;
use comfy_table::{Cell, Color, ContentArrangement, Table, presets::UTF8_BORDERS_ONLY};

use super::{CommandResult, EXIT_ERROR, load_config_or_exit, render_options};

pub(super) async fn run(command: HistoryCommand) -> CommandResult {
    let config = load_config_or_exit(config_path(&command))?;
    let store = SqliteHistoryStore::connect(&config.database.url)
        .await
        .map_err(|error| {
            eprintln!("Error: {error}");
            EXIT_ERROR
        })?;

    match command {
        HistoryCommand::List {
            output, verbose, ..
        } => list(&store, output, verbose).await,
        HistoryCommand::Show {
            run_id,
            output,
            verbose,
            ..
        } => show(&store, &run_id, output, verbose).await,
    }
}

fn config_path(command: &HistoryCommand) -> Option<&Path> {
    match command {
        HistoryCommand::List { config, .. } | HistoryCommand::Show { config, .. } => {
            config.as_deref()
        }
    }
}

async fn list(store: &SqliteHistoryStore, output: Format, verbose: bool) -> CommandResult {
    let runs = store.list().await.map_err(|error| {
        eprintln!("Error: {error}");
        EXIT_ERROR
    })?;

    match output {
        Format::Json => match serde_json::to_string(&runs) {
            Ok(rendered) => println!("{rendered}"),
            Err(error) => {
                eprintln!("Error: {error}");
                return Err(EXIT_ERROR);
            }
        },
        Format::Text if runs.is_empty() => println!("No saved diagnosis runs found."),
        Format::Text => print_table(&runs, verbose),
    }
    Ok(())
}

async fn show(
    store: &SqliteHistoryStore,
    run_id: &str,
    output: Format,
    verbose: bool,
) -> CommandResult {
    let result = store.get(run_id).await.map_err(|error| {
        eprintln!("Error: {error}");
        EXIT_ERROR
    })?;
    let Some(result) = result else {
        eprintln!("Error: run {run_id} not found.");
        return Err(EXIT_ERROR);
    };

    let report = Report::new(result);
    match output {
        Format::Text => print!("{}", text::render(&report, &render_options(verbose))),
        Format::Json => match serde_json::to_string_pretty(&json::render(&report, verbose)) {
            Ok(rendered) => println!("{rendered}"),
            Err(error) => {
                eprintln!("Error: {error}");
                return Err(EXIT_ERROR);
            }
        },
    }
    Ok(())
}

fn print_table(runs: &[RunSummary], verbose: bool) {
    let color = std::io::stdout().is_terminal();
    let mut table = Table::new();
    table
        .load_preset(UTF8_BORDERS_ONLY)
        .set_content_arrangement(ContentArrangement::Disabled);

    let mut header = vec!["Run ID", "Time", "Model"];
    if verbose {
        header.extend(["Source", "Engine", "Target"]);
    }
    header.extend(["Health", "Fired"]);
    table.set_header(header);

    for run in runs {
        let saved = run.saved_at.format("%Y-%m-%d %H:%M").to_string();
        let model = run.model_name.clone().unwrap_or_else(|| "—".to_string());
        let mut row = vec![Cell::new(run.run_id), Cell::new(saved), Cell::new(model)];
        if verbose {
            row.push(Cell::new(run.metrics_source.to_string()));
            row.push(Cell::new(run.engine.to_string()));
            row.push(Cell::new(run.target_id.as_deref().unwrap_or("—")));
        }
        let health = Cell::new(run.health.to_string());
        row.push(if color {
            health.fg(health_color(run.health))
        } else {
            health
        });
        row.push(Cell::new(run.fired_count));
        table.add_row(row);
    }

    println!("{table}");
}

fn health_color(health: Health) -> Color {
    match health {
        Health::Ok => Color::Green,
        Health::Info => Color::Blue,
        Health::Warning => Color::Yellow,
        Health::Critical => Color::Red,
    }
}
